### PR TITLE
autoconfigure k8s namespace from filesystem

### PR DESCRIPTION
**Pull request by bparees**

### CODE REVIEW

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java [hunk 1]
```diff
@@ -101,6 +109,7 @@ public KubernetesCloud(String name, List<? extends PodTemplate> templates, Strin
 
         this.serverUrl = serverUrl;
         this.namespace = namespace;
+
```
Comment (bparees) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]: note: i really wanted to just put the getNamespace() call here, but this constructor didn't seem to be getting called based on the behavior i was observing.  i'm not sure how that's possible, so if someone can point me to a better way to initialize namespace from the filesystem (when namespace is not explicitly set in the config) i'd be happy to do that instead.

**Comment by bparees** [comment]
@jimmidyson @spadgett fyi.

**Comment by bparees** [comment]
@carlossg what's the process for moving forward w/ reviewing this change?

**Comment by carlossg** [comment]
Can you reformat using Java conventions? Otherwise diffs get harder to read

**Comment by bparees** [comment]
@carlossg happy to reformat it to anything you prefer, but not sure what qualifies as "Java conventions" these days.  I used default eclipse formatting and import ordering.  Do you have an eclipse format I can import and use?

**Comment by carlossg** [comment]
You can switch to `Java Conventions` in http://help.eclipse.org/mars/index.jsp?topic=%2Forg.eclipse.jdt.doc.user%2Freference%2Fpreferences%2Fjava%2Fcodestyle%2Fref-preferences-formatter.htm
or in a per-project basis

I'll add some comments about the code too

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java [hunk 2]
```diff
@@ -230,10 +242,11 @@ private Pod getPodTemplate(KubernetesSlave slave, Label label) {
         url = url.endsWith("/") ? url : url + "/";
         env.add(new EnvVar("JENKINS_JNLP_URL", url + slave.getComputer().getUrl() + "slave-agent.jnlp", null));
 
-        for (PodEnvVar podEnvVar :template.getEnvVars()) {
-            env.add(new EnvVar(podEnvVar.getKey(), podEnvVar.getValue(), null));
+        if(template.getEnvVars()!=null) {
+	        for (PodEnvVar podEnvVar :template.getEnvVars()) {
+	            env.add(new EnvVar(podEnvVar.getKey(), podEnvVar.getValue(), null));
+	        }
```
Comment (carlossg) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]: better if this goes into its separate commit as it is unrelated to the main commit

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java [hunk 3]
```diff
@@ -151,7 +160,10 @@ public void setSkipTlsVerify(boolean skipTlsVerify) {
     }
 
     public String getNamespace() {
-        return namespace;
+        if(namespace==null || "".equals(namespace)) {
```
Comment (carlossg) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]: StringUtils.isEmpty(namespace) will take care of both (not important, just FYI)

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/ConfigUtils.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class ConfigUtils {
+	private static final Logger LOGGER = Logger.getLogger(ConfigUtils.class.getName());
+	public static final String KUBERNETES_NAMESPACE_PATH = "/var/run/secrets/kubernetes.io/serviceaccount/namespace";
+
+	public static String getNamespace(String namespace) {
```
Comment (carlossg) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/ConfigUtils.java]: the parameter is always null or "", so I don't think it's needed at all.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/ConfigUtils.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+		LOGGER.log(Level.FINEST,
+				"getNamespace called with "+namespace);
+		if (namespace != null && !namespace.equals("")) {
```
Comment (carlossg) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/ConfigUtils.java]: this `if` shouldn't be here or the filesystem namespace is never read when a default namespace is passed

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java (hunk 1 above)
Comment (carlossg) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]: IIUC this gets called when you save the plugin configuration. When the configuration is already saved then readResolve and methods marked with DataBoundSetter are called

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/ConfigUtils.java (hunk 5 above)
Comment (bparees) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/ConfigUtils.java]: yeah the intent was to be able to blindly call this if the calling code didn't know if it already had a namespace or not.  If it already had a valid value, we'd short-circuit looking at the filesystem for a value.

I may revisit that pattern as it's a bit ugly.  But it was intentional.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java (hunk 1 above)
Comment (bparees) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]: ok, then i guess the getSystemNamespace() call wouldn't belong here anyway.  thanks!

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java (hunk 2 above)
Comment (bparees) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]: will do.

**Comment by bparees** [comment]
@carlossg updated based on your comments (and commit split out), note that when i tried to format the KubernetesCloud.java file, it introduced a bunch of changes, so i did not format that file.  But the new file has been formatted w/ the Java Conventions.

**Comment by jimmidyson** [comment]
Another way to do this is to upgrade the kubernetes-client dependency which should do this automatically for you, while also taking advantage of other bug fixes & improvements.

**Comment by bparees** [comment]
@jimmidyson i built a local version of this plugin w/ the kubernetes client version bumped to 1.3.97 but i saw no change in the namespace behavior...  are there changes required in this plugin to take advantage of the auto-namespacing of the kubernetes client?  what exactly does it do?

**Comment by jimmidyson** [comment]
@bparees I had a quick look & it looks like there are a number of things in this plugin which override that default behaviour. Would require a bit more work to remove all those things from this plugin so they only override default behaviour if set.

**Comment by bparees** [comment]
@jimmidyson ok, with that in mind maybe we can just stick w/ the existing implementation.
@carlossg any opinion?

**Comment by bparees** [comment]
for what it's worth i've worked around this issue by writing a startup script that hacks the config.xml to inject the right namespace value based on where the container is running, but i still would like to see this issue addressed either via this PR or another mechanism.

**Comment by carlossg** [comment]
@jimmidyson something like 1e3f701 to leave values as null for the kuebernetes-client to use autodiscovery?

**Comment by iocanel** [comment]
@carlossg: This will result in IllegalArgumentException. when we try to create the pod.

We should go the other way around. Pass the namespace if explicitly specified in the client config and not specify a client (let the client fallback to the 'current' namespace)  when we use the client.

**Comment by iocanel** [comment]
Let me create a pr for this.

**Comment by bparees** [comment]
@iocanel @carlossg @jimmidyson i'm going to close this one since it doesn't look like the direction we're going.  Thanks for picking it up.

**Comment by iocanel** [comment]
@bparees: my comment was about the implementation (and more specificallyhttps://github.com/jenkinsci/kubernetes-plugin/commit/1e3f701949530faca16257df977f69be2d1e4826) and not the overall direction.
